components/script/dom/response.rs

```diff
@@ -179,64 +179,40 @@ impl Response {
         Ok(r)
     }
 
-    // https://fetch.spec.whatwg.org/#concept-body-consume-body
-    #[allow(unrooted_must_root)]
-    fn consume_body(&self, body_type: BodyType) -> Fallible<Rc<Promise>> {
-        let promise = Promise::new(self.global().r());
-
-        // Step 1
-        if self.BodyUsed() || self.locked() {
-            promise.maybe_reject_error(promise.global().r().get_cx(), Error::Type(
-                "The response's stream is disturbed or locked".to_string()));
-            return Ok(promise);
-        }
-
-        // Steps 2-4
-        // TODO: Body does not yet have a stream.
-
-        // Step 5
-        let pkg_data_results: String = try!(self.run_package_data_algorithm(body_type));
-        let pkg_data_results: USVString = USVString(pkg_data_results);
-        promise.maybe_resolve_native(promise.global().r().get_cx(), &pkg_data_results);
-        Ok(promise)
-    }
-
     // https://fetch.spec.whatwg.org/#concept-body-locked
-    fn locked(&self) -> bool {
+    pub fn locked(&self) -> bool {
         // TODO: ReadableStream is unimplemented. Just return false
         // for now.
         false
     }
+}
 
-    // https://fetch.spec.whatwg.org/#concept-body-package-data
-    fn run_package_data_algorithm(&self, body_type: BodyType) -> Fallible<String> {
-        match body_type {
-            BodyType::Text => {
-                // return result of running utf-8 decode here
-                // using encoding crate on all bytes instead of
-                // individually processing each token
-                // TODO: also check NetTraitsResponseBody::Receiving(bytes) = *self.body.borrow()
-                if let NetTraitsResponseBody::Done(ref bytes) = *self.body.borrow() {
-                    self.body_used.set(true);
-                    let result: Result<String, Error> =
-                        UTF_8.decode(bytes, DecoderTrap::Replace).map_err(|e| Error::Type((&*e).to_owned()));
-                    *self.body.borrow_mut() = NetTraitsResponseBody::Empty;
-                    return result;
-                }
-                return Ok("".to_owned());
-            },
-            _ => Err(Error::Type(
-                "Unable to process bytes".to_string()))
+impl BodyTrait for Response {
+    fn get_body_used_trait(&self) -> bool {
+        self.BodyUsed()
+    }
+
+    fn locked_trait(&self) -> bool {
+        self.locked()
+    }
+
+    fn take_body(&self) -> Option<Vec<u8>> {
+        let body: NetTraitsResponseBody = mem::replace(&mut *self.body.borrow_mut(), NetTraitsResponseBody::Empty);
+        if let NetTraitsResponseBody::Done(bytes) = body {
```

Comment:
Let's use this to avoid duplication:
```rust
match body {
    NetTraitsResponseBody::Done(bytes) | NetTraitsResponseBody::Receiving(bytes) => {
        self.body_used.set(true);
        Some(bytes)
    }
    _ => None,
}
```